include/openssl/err.h

```diff
@@ -163,14 +163,43 @@ struct err_state_st {
 #  define X509err(f, r) ERR_raise_data(ERR_LIB_X509, (r), NULL)
 # endif
 
-# define ERR_PACK(l,f,r) ( \
-        (((unsigned int)(l) & 0x0FF) << 24L) | \
-        (((unsigned int)(f) & 0xFFF) << 12L) | \
-        (((unsigned int)(r) & 0xFFF)       ) )
-# define ERR_GET_LIB(l)          (int)(((l) >> 24L) & 0x0FFL)
-# define ERR_GET_FUNC(l)         (int)(((l) >> 12L) & 0xFFFL)
-# define ERR_GET_REASON(l)       (int)( (l)         & 0xFFFL)
-# define ERR_FATAL_ERROR(l)      (int)( (l)         & ERR_R_FATAL)
+/*
+ * The error code currently packs as follows (viewed as hex nibbles):
+ *
+ * LL rRRRRR
+ *
+ * Where LL is the library code, r is the reason flags, and rRRRRR is the
+ * reason code.
+ * Do note that the reason flags is part of the reason code, and could as
+ * well be seen as a section of all possible reason codes.  We do this for
+ * backward compatibility reasons, i.e. how ERR_R_FATAL was implemented.
+ *
+ * System errors (ERR_LIB_SYS) are structured the same way, except they
+ * don't have any reason flag.
+ *
+ * LL RRRRRR
```

Comment:
so this is special encoding for a specific LL value; maybe put that value here?
And consider making ```ERR_LIB_SYS``` by define'd to zero.

Reply:
Using zero is difficult, as that value is used internally to reach ERR_R_ reasons.  Happens here:

https://github.com/openssl/openssl/blob/94941cada25433a7dca35b5b9f8cbb751ab65ab3/crypto/err/err.c#L649-L652

So, using zero for the lib part in recorded system errors as well comes with a (very) likely risk of clashes, and misleading error text.

Reply:
Actually, I have other thoughts on system errors, see https://github.com/openssl/openssl/pull/12314#issuecomment-652801232